### PR TITLE
Return finer grain errors in libzfs unmount_one

### PR DESCRIPTION
**Pull request by don-brady**
### Motivation and Context
For zfs CLI users and libzfs clients of `zfs_unmount()`,  a failed mount reports a generic "_umount failed, cannot unmount_" error.  Ideally a more specific error, like "_dataset is busy_" or "_permission denied_" could be returned.

The underlying posix `umount(2)` system call returns more specific errors that are currently all mapped to `EZFS_UMOUNTFAILED`.

### Description
Added errno mappings to unmount_one() in libzfs.

~~Also for the Linux implementation of `do_unmount()` made the process exec version follow the same error protocol as the `umount(2)` system call:  on error, -1 is returned, and `errno` is set appropriately.~~

Changed `do_unmount()` implementation to return errno errors directly like is done for `do_mount()` and others.

### How Has This Been Tested?
**ZTS** -- ran mount related tests:
- [x] functional/mount
- [x] functional/cli_root/zfs_mount
- [x] functional/cli_root/zfs_unmount
	
**CLI** -- confirmed results for failed mounts
_Before_
```
$ zfs unmount wool/parent
cannot unmount '/wool/parent': unmount failed
```
_After_ (2 examples)
```
$ zfs unmount wool/parent
cannot unmount '/wool/parent': permission denied

$ sudo zfs unmount wool/parent
cannot unmount '/wool/parent/child': pool or dataset is busy
```

**Libzfs consumer** -- confirmed results from `libzfs_errno()` and  `libzfs_error_description()`
_Before_
```
ZfsDataServiceImpl$UnmountFailedException: UMOUNTFAILED: umount failed, cannot unmount '/domain0/group-2/oracle_db_container-220243/oracle_timeflow-220245/datafile'
```
_After_
```
ZfsDataServiceImpl$UnmountFailedException: BUSY: pool or dataset is busy, cannot unmount '/domain0/group-2/oracle_db_container-6/oracle_timeflow-7/datafile'
```

### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: lib/libzfs/libzfs_mount.c [hunk 1]
```diff
@@ -552,8 +552,30 @@ unmount_one(libzfs_handle_t *hdl, const char *mountpoint, int flags)
 	int error;
 
 	error = do_unmount(mountpoint, flags);
+	/* On error, -1 is returned, and errno is set appropriately */
```
Comment (behlendorf) [inline lib/libzfs/libzfs_mount.c:555]: It feels a little dodgy to depend on `errno` still being set correctly here.  Perhaps we should update `do_unmount` to actually return the `errno` set by the syscall.  This way unrelated changes to `do_unmount()` can't accidentally break this logic.

Comment (don-brady) [reply lib/libzfs/libzfs_mount.c:555]: Good point!  I had thought about making `do_unmount` be more like similar functions that return actual error.  I'll make the change.

File: lib/libzfs/os/linux/libzfs_mount_os.c [hunk 2]
```diff
@@ -377,7 +377,9 @@ int
 do_unmount(const char *mntpt, int flags)
 {
 	if (!libzfs_envvar_is_set("ZFS_MOUNT_HELPER")) {
-		return (umount2(mntpt, flags));
+		int rv = umount2(mntpt, flags);
+
+		return (rv < 0 ? errno : rv);
```
Comment (mmaybee) [inline lib/libzfs/os/linux/libzfs_mount_os.c:382]: I think it would be slightly more clear to use:
`return (rv < 0 ? errno : 0);`

**Review by mmaybee — CHANGES_REQUESTED** [review]

File: lib/libzfs/os/freebsd/libzfs_zmount.c [hunk 3]
```diff
@@ -120,9 +120,13 @@ do_mount_(const char *spec, const char *dir, int mflag, char *fstype,
 int
 do_mount(zfs_handle_t *zhp, const char *mntpt, char *opts, int flags)
 {
+	int rv;
 
-	return (do_mount_(zfs_get_name(zhp), mntpt, flags, MNTTYPE_ZFS, NULL, 0,
-	    opts, sizeof (mntpt)));
+	rv = do_mount_(zfs_get_name(zhp), mntpt, flags, MNTTYPE_ZFS, NULL, 0,
+	    opts, sizeof (mntpt));
+	if (rv < 0)
```
[the remark this answers is not in the document]
Comment (don-brady) [reply lib/libzfs/os/freebsd/libzfs_zmount.c]: Ah, great catch!  The change should have been in ` do_unmount()` which is making a syscall and sets errno.  Fixed in latest commit.

File: lib/libzfs/os/linux/libzfs_mount_os.c (hunk 2 above)
Comment (don-brady) [reply lib/libzfs/os/linux/libzfs_mount_os.c]: fixed

File: lib/libzfs/os/freebsd/libzfs_zmount.c [hunk 4]
```diff
@@ -120,16 +120,19 @@ do_mount_(const char *spec, const char *dir, int mflag, char *fstype,
...
 
 int
 do_unmount(const char *mntpt, int flags)
 {
+	int rv;
 
-	return (unmount(mntpt, flags));
+	rv = unmount(mntpt, flags);
+	if (rv < 0)
+		return (errno);
```
Comment (mmaybee) [inline lib/libzfs/os/freebsd/libzfs_zmount.c:134]: This could be simplified:

`if (unmount(mntpt, flags) < 0)
            return (errno);
return (0);`

**Review by mmaybee — APPROVED** [review]

File: lib/libzfs/os/freebsd/libzfs_zmount.c (hunk 4 above)
Comment (don-brady) [reply lib/libzfs/os/freebsd/libzfs_zmount.c:132]: indeed! fixed

**Review by behlendorf — APPROVED** [review]
